exonum/src/runtime/rust/api.rs

```diff
@@ -160,10 +162,136 @@ impl ServiceApiScope {
         self
     }
 
+    /// Same as `endpoint`, but also add a warning about this endpoint being deprecated to the response.
+    pub fn deprecated_endpoint<Q, I, F, R>(
```

Comment:
Is it possible to use the following decorator-based approach? (Not sure about the accuracy of the "decorator" term.)

```rust
struct ServiceApi;

impl ServiceApi {
    fn some_handler(
        state: &api::ServiceApiState<'_>,
        arg: SomeRequest,
    ) -> api::Result<SomeResponse> {
        // ...
    }
}

let scope: &mut ServiceApiScope = ...;
scope.endpoint(Deprecated::new(ServiceApi::some_handler).with_date(Utc.ymd(2020, 06, 20))
```

Would minimize the amount of abstractions and could allow to introduce new "decorators" in the future.

Reply:
Well, it seems that currently implementing this feature on the `rust` runtime API level will require significant architecture changes, so I suggest to left things as-is *for now*, but return to this question once we will face the need in this functionality for at least one more thing.